### PR TITLE
Use only fs.NewDownloader for file://.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Changes the upgrade artifacts fetcher to only use the `fs.NewDownloader` when a `--source-uri` is passed with `file://` prefix.

## Why is it important?

Ensures that when a `file://` source URI is provided that a file will never be downloaded from the artifacts API. Upgrade requested to use a local file it should only look there.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test **(already used in tests)**

## How to test this PR locally

```
$ elastic-agent upgrade --skip-verify --source-uri file:///home/doesnt-exist 8.12.0-SNAPSHOT
Error: Failed trigger upgrade of daemon: failed download of agent binary: unable to download package: package '/home/ubuntu/elastic-agent-8.12.0-SNAPSHOT-linux-arm64.tar.gz' not found: open /home/ubuntu/elastic-agent-8.12.0-SNAPSHOT-linux-arm64.tar.gz: permission denied
For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.12/fleet-troubleshooting.html
```

### CODE REVIEW

**Review by michalpristas — APPROVED** [review]
looks good

File: internal/pkg/agent/application/upgrade/step_download.go [hunk 1]
```diff
@@ -46,32 +50,63 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
...
 			// path instead of looking into the installed downloads directory
 			settings.DropPath = strings.TrimPrefix(sourceURI, "file://")
+
+			// use specific function that doesn't perform retries on download as its
+			// local and no retry should be performed
+			downloaderFunc = u.downloadOnce
+
+			// set specific downloader, local file just uses the fs.NewDownloader
+			// no fallback is allowed because it was requested that this specific source be used
+			factory = func(ver *agtversion.ParsedSemVer, l *logger.Logger, config *artifact.Config, d *details.Details) (download.Downloader, error) {
```
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/step_download.go:75]: you can even move this to fs and have `factory = fs.NewDownloaderFactory` to make it a bit more readable

Comment (blakerouse) [reply internal/pkg/agent/application/upgrade/step_download.go:75]: i didn't do that as the factory interface is private, takes arguments that the download module doesn't really need to import as its not used

**Comment by blakerouse** [comment]
buildkite test this

**Comment by cmacknz** [comment]
Buildkite failure is only in the merge coverage reports step, force merging.

**Comment by ycombinator** [comment]
@blakerouse would you mind backporting this PR to `7.17`, per our discussion in the team meeting?  Thanks.

**Pull request merged**